go-controller/pkg/node/gateway_localnet.go

```diff
@@ -270,57 +444,78 @@ func localnetIptRules(svc *kapi.Service, gatewayIP string) []iptRule {
 }
 
 type localnetNodePortWatcherData struct {
-	ipt       util.IPTablesHelper
-	gatewayIP string
+	localNetdata []*localnetData
 }
 
 func (npw *localnetNodePortWatcherData) addService(svc *kapi.Service) error {
 	if !util.ServiceTypeHasNodePort(svc) {
 		return nil
 	}
-	rules := localnetIptRules(svc, npw.gatewayIP)
+	lnData := getLocalNetDataForService(npw.localNetdata, svc)
+	rules := localnetIptRules(svc, lnData.gatewayIP.String())
 	klog.V(5).Infof("Add rules %v for service %v", rules, svc.Name)
-	return addIptRules(npw.ipt, rules)
+	return addIptRules(lnData.ipt, rules)
 }
 
 func (npw *localnetNodePortWatcherData) deleteService(svc *kapi.Service) error {
 	if !util.ServiceTypeHasNodePort(svc) {
 		return nil
 	}
-	rules := localnetIptRules(svc, npw.gatewayIP)
+	lnData := getLocalNetDataForService(npw.localNetdata, svc)
+	rules := localnetIptRules(svc, lnData.gatewayIP.String())
 	klog.V(5).Infof("Delete rules %v for service %v", rules, svc.Name)
-	delIptRules(npw.ipt, rules)
+	delIptRules(lnData.ipt, rules)
 	return nil
 }
 
-func localnetNodePortWatcher(ipt util.IPTablesHelper, wf *factory.WatchFactory, gatewayIP net.IP) error {
-	// delete all the existing OVN-NODEPORT rules
+func getLocalNetDataForService(localNetdata []*localnetData, svc *kapi.Service) *localnetData {
+	var ipVersion = ipv4
+	if utilnet.IsIPv6String(svc.Spec.ClusterIP) {
+		ipVersion = ipv6
+	}
+	for _, lnData := range localNetdata {
+		if lnData.ipVersion == ipVersion {
+			return lnData
```

Comment:
it seems like it would be simpler to just have

    type localnetNodePortWatcherData struct {
            ipv4Data *localnetData
            ipv6Data *localnetData
    }

Reply:
To me it's actually a bit simpler the way I implemented . Because based on the configuration either one OR both have a valid value and the other one nil. Also we need only one of them for nodeport service but need both of them for base rules. Hence having them in a slice makes it easier for all the scenarios. Just loop through them  and use the right one based on context throughout.  I would prefer to keep it this way.